nomad/state/state_store.go

```diff
@@ -2540,6 +2540,19 @@ func (s *StateStore) UpsertEvalsTxn(index uint64, evals []*structs.Evaluation, t
 	return nil
 }
 
+// ErrDuplicateEval indicates that the eval is already present in state and ought not
+// to be reprocessed again.
+//
+// Such duplication might occur in the background compatibility path in job registration
+// or deregistration
+var ErrDuplicateEval = fmt.Errorf("eval already exists")
+
+func isPotentialDuplicateEval(eval *structs.Evaluation) bool {
+	return eval.Status == structs.EvalStatusPending &&
+		(eval.TriggeredBy == structs.EvalTriggerJobRegister ||
```

Comment:
It looks like we hit `nestedUpsertEval` in the deregister case as well: should we also allow for `eval.TriggeredBy == structs.EvalTriggerJobDeregister` here?

Reply:
Good catch!  Thanks